cranelift/fuzzgen/src/function_generator.rs

```diff
@@ -0,0 +1,253 @@
+
+        for _ in 0..self.u.int_in_range(self.config.signature_rets.clone())? {
+            sig.returns.push(self.generate_abi_param()?);
+        }
+
+        Ok(sig)
+    }
+
+    /// Creates a new var
+    fn create_var(&mut self, builder: &mut FunctionBuilder, ty: Type) -> Result<Variable> {
+        let id = self.vars.len();
+        let var = Variable::new(id);
+        builder.declare_var(var, ty);
+        self.vars.push((ty, var));
+        Ok(var)
+    }
+
+    fn vars_of_type(&self, ty: Type) -> Vec<Variable> {
+        self.vars
+            .iter()
+            .filter(|(var_ty, _)| *var_ty == ty)
+            .map(|(_, v)| *v)
+            .collect()
+    }
+
+    /// Get a variable of type `ty` from the current function
+    fn get_variable_of_type(&mut self, ty: Type) -> Result<Variable> {
+        let opts = self.vars_of_type(ty);
+        let var = self.u.choose(&opts[..])?;
+        Ok(*var)
+    }
+
+    /// Generates an instruction(`iconst`/`fconst`/etc...) to introduce a constant value
+    fn generate_const(&mut self, builder: &mut FunctionBuilder, ty: Type) -> Result<Value> {
+        let imm64 = match ty {
+            I8 => self.u.arbitrary::<i8>()? as i64,
+            I16 => self.u.arbitrary::<i16>()? as i64,
+            I32 => self.u.arbitrary::<i32>()? as i64,
+            I64 => self.u.arbitrary::<i64>()?,
+            _ => unreachable!(),
+        };
+        let val = builder.ins().iconst(ty, imm64);
+
+        Ok(val)
+    }
+
+    fn generate_return(&mut self, builder: &mut FunctionBuilder) -> Result<()> {
+        let ret_params = builder.func.signature.returns.clone();
+
+        let vars = ret_params
+            .iter()
+            .map(|p| self.get_variable_of_type(p.value_type))
+            .collect::<Result<Vec<_>>>()?;
+
+        let vals = vars
+            .into_iter()
+            .map(|v| builder.use_var(v))
+            .collect::<Vec<_>>();
+
+        builder.ins().return_(&vals[..]);
+        Ok(())
+    }
+
+    /// Inserts a random instruction into the block
+    fn generate_instruction(&mut self, builder: &mut FunctionBuilder) -> Result<()> {
+        let (op, args, rets, inserter) = *self.u.choose(OPCODE_SIGNATURES)?;
+        inserter(self, builder, op, args, rets)
+    }
+
+    pub fn generate(mut self) -> Result<Function> {
+        let sig = self.generate_signature()?;
+
+        let mut fn_builder_ctx = FunctionBuilderContext::new();
+        let mut func = Function::with_name_signature(ExternalName::user(0, 0), sig.clone());
+
+        let mut builder = FunctionBuilder::new(&mut func, &mut fn_builder_ctx);
+        let block0 = builder.create_block();
+        builder.append_block_params_for_function_params(block0);
+        builder.switch_to_block(block0);
+        builder.seal_block(block0);
+
+        // Define variables for the function signature
+        for (i, param) in sig.params.iter().enumerate() {
+            let var = self.create_var(&mut builder, param.value_type)?;
```

Comment:
Now that I'm looking at this again, I don't think its quite right, we probably should be assigning the block params to the pool of variables instead of creating new ones.

I'll look at this again tommorow

Reply:
I think this should be right? At least, my understanding is that the blockparams are SSA values while the variables are a higher-level abstraction; it would be more confusing to have a pool of both and then selectively lower vars to SSA values or use an SSA value directly, I think.

Reply:
Ah, this works only because the first block is special and has the function params. I think for the other blocks we have to associate the block params to existing variables? Otherwise we run into the issue you were pointing out earlier where we may not have defined a variable depending on where we came from.

Anyway, I'll probably start working on the control flow improvements today, and should be in a more informed position to address this when I submit the next PR.